src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -626,17 +631,23 @@ private struct UrlDecoder
             private void FlushBytes()
             {
                 Debug.Assert(_numBytes > 0);
+                if (_charBuffer == null)
+                    _charBuffer = new char[_bufferSize];
+
                 _numChars += _encoding.GetChars(_byteBuffer, 0, _numBytes, _charBuffer, _numChars);
                 _numBytes = 0;
             }
 
             internal UrlDecoder(int bufferSize, Encoding encoding)
             {
+                _containsUnsafe = false;
+                _containsSpaces = false;
+
                 _bufferSize = bufferSize;
                 _encoding = encoding;
 
-                _charBuffer = new char[bufferSize];
-                
+                _charBuffer = null; // char buffer created on demand
```

Comment:
Is this really useful?  Empty string is already special cased to avoid decoding.  So isn't this only valuable in the case where all of the inputs are bytes rather than chars, and to enable that you're then doing an extra null check branch on every AddChar?  Doesn't seem like a good tradeoff.